src/core/model/Stroke.cpp

```diff
@@ -424,18 +642,76 @@ void Stroke::calcSize() const {
     Element::snappedBounds = Rectangle<double>(minSnapX, minSnapY, maxSnapX - minSnapX, maxSnapY - minSnapY);
 }
 
-auto Stroke::getErasable() -> ErasableStroke* { return this->eraseable; }
+auto Stroke::getErasable() const -> ErasableStroke* { return this->erasable; }
 
-void Stroke::setErasable(ErasableStroke* eraseable) { this->eraseable = eraseable; }
+void Stroke::setErasable(ErasableStroke* erasable) { this->erasable = erasable; }
 
 auto Stroke::getStrokeCapStyle() const -> StrokeCapStyle { return this->capStyle; }
 
 void Stroke::setStrokeCapStyle(const StrokeCapStyle capStyle) { this->capStyle = capStyle; }
 
-void Stroke::debugPrint() {
+void Stroke::debugPrint() const {
     g_message("%s", FC(FORMAT_STR("Stroke {1} / hasPressure() = {2}") % (uint64_t)this % this->hasPressure()));
 
-    for (auto&& p: points) { g_message("%lf / %lf", p.x, p.y); }
+    for (auto&& p: points) { g_message("%lf / %lf / %lf", p.x, p.y, p.z); }
 
     g_message("\n");
 }
+
+std::optional<Interval<double>> Stroke::intersectLineWithRectangle(const Point& p, const Point& q,
+                                                                   const Rectangle<double>& rectangle) {
+    if (p.x == q.x) {
+        // Vertical segment
+        double x = p.x;
+        if (p.y == q.y || x <= rectangle.x || x >= rectangle.x + rectangle.width) {
+            return std::nullopt;
+        }
+        double norm = 1.0 / (q.y - p.y);
+        double tUp = (rectangle.y - p.y) * norm;
+        double tBottom = tUp + rectangle.height * norm;
+        return Interval<double>::getInterval(tUp, tBottom);
+    }
+
+    if (p.y == q.y) {
+        // Horizontal segment
+        double y = p.y;
+        if (y <= rectangle.y || y >= rectangle.y + rectangle.height) {
+            return std::nullopt;
+        }
+        double norm = 1.0 / (q.x - p.x);
+        double tUp = (rectangle.x - p.x) * norm;
+        double tBottom = tUp + rectangle.width * norm;
+        return Interval<double>::getInterval(tUp, tBottom);
+    }
```

Comment:
First, this is duplicate code and should be moved into a more generic lambda.
Second, I don't think, that both branches are too unlikely to specify a specific function for them.
The generic case might be enough.

Reply:
No, those cases do happen as soon as the rectangle tool is used, or the pdf ruler, or a line is recognized and snapped to the grid, etc.
Then, you're right, it could be made into a lambda or so.

Reply:
done